lib/config.js

```diff
@@ -360,21 +379,22 @@ function parseConfig (configFilePath, cliOptions) {
         configModule = configModule.default
       }
     } catch (e) {
-      log.error('Error in config file!\n  ' + e.stack || e)
+      const errorMessageSegments = []
+      errorMessageSegments.push('Error in config file!\n  ' + e.stack || e)
 
       const extension = path.extname(configFilePath)
       if (extension === '.coffee' && !COFFEE_SCRIPT_AVAILABLE) {
-        log.error('You need to install CoffeeScript.\n  npm install coffeescript --save-dev')
+        errorMessageSegments.push('You need to install CoffeeScript.\n  npm install coffeescript --save-dev')
       } else if (extension === '.ls' && !LIVE_SCRIPT_AVAILABLE) {
-        log.error('You need to install LiveScript.\n  npm install LiveScript --save-dev')
+        errorMessageSegments.push('You need to install LiveScript.\n  npm install LiveScript --save-dev')
       } else if (extension === '.ts' && !TYPE_SCRIPT_AVAILABLE) {
-        log.error('You need to install TypeScript.\n  npm install typescript ts-node --save-dev')
+        errorMessageSegments.push('You need to install TypeScript.\n  npm install typescript ts-node --save-dev')
       }
-      return process.exit(1)
+      const configModuleImportErrorMessage = errorMessageSegments.join('\n')
+      return fail(configModuleImportErrorMessage)
```

Comment:
I think "You need to install ..." logs are more of a hint on where the error could have come from, so we can skip them from the thrown error and only keep them as regular logs. With this assumption, we can move the actual error log and `process.exit(1)` together and avoid `errorMessageSegments`.

```suggestion
      return fail('Error in config file!\n  ' + e.stack || e)
```

Reply:
> Please also update the usage of `parseConfig` in Karma's source code, so it does not rely on the deprecated behavior (last point in [#3631 (comment)](https://github.com/karma-runner/karma/issues/3631#issuecomment-767006344)).

```javascript
    let config
    try {
      // A)
      // Did you want this and an option passed to it as the third argument?
      config = cfg.parseConfig(cliOptions.configFile, cliOptions)

      // B)
      // or did you want the new behavior enabled immediatly since the catch
      // block will done or exit anyway?
      config = cfg.parseConfig(cliOptions.configFile, cliOptions, { throwErrors: true })
    } catch (parseConfigError) {
      (done || process.exit)(1)
    }
```

Did you want the server code to enable the new behavior?

---

> ... we can move the actual error log and process.exit(1) together and get rid of errorMessageSegments variable.

> And add a unit test for `parseConfig` function with `throwErrors: true`.

Will do!

Reply:
B), so we behave the same as before (exit the process), but no longer rely on the deprecated `process.exit()` call in the `parseConfig` method and don't need to change our code when we remove it in the next major release.

Two notes:
- We should probably `return` in the catch block to prevent any further execution in case custom `done` is passed in.
- It may be cleaner to use a default argument instead of inline fallback: `constructor (cliOptions, done = process.exit)`.

Reply:
> * It may be cleaner to use a default argument instead of inline fallback: `constructor (cliOptions, done = process.exit)`.

The previous behavior was "if `done` is **falsey**, then call process.exit()."

`constructor (cliOptions, done = process.exit)` changes it to "if `done` is **`undefined`**, then call process.exit()."

**Are you ok with this?**

```javascript
  constructor (cliOptions, done = process.exit) {
    super()
    logger.setupFromConfig(cliOptions)

    this.log = logger.create('karma-server')

    this.loadErrors = []

    let config
    try {
      config = cfg.parseConfig(cliOptions.configFile, cliOptions, { throwErrors: true })
    } catch (parseConfigError) {
      done(1)
      return
    }

    this.log.debug('Final config', util.inspect(config, false, /** depth **/ null))

    let modules = [{
      helper: ['value', helper],
      logger: ['value', logger],
      done: ['value', done], // PREVIOUSLY:    done: ['value', done || process.exit],
```

---

Neither case guards against `done` being a non-callable/non-function value. That kind of change is probably out of scope for this PR, and that is if you determine it to be necessary at all.

Reply:
You're right. Technically falling back on `undefined` instead of a falsy value is a breaking change. So let's play it safe and restore the original behavior with the TODO to change it in the next major.

Besides this, it LGTM! Thanks again for doing this work :+1:

----

> Neither case guards against done being a non-callable/non-function value. That kind of change is probably out of scope for this PR, and that is if you determine it to be necessary at all.

Yeah, it's out of scope for this PR and I doubt it's worth the effort in general.

Reply:
With that in mind, should I make the following change (Option A) as well?

```diff
     let config
     try {
       config = cfg.parseConfig(cliOptions.configFile, cliOptions, { throwErrors: true })
     } catch (parseConfigError) {
-      done(1)
+      process.exit(1) // ------ // Option A
+      (done || process.exit)(1) // Option B
       return
     }
```
Previously, `parseConfig` only exited the process here. It did not attempt to call `done`.

Reply:
I think (B) is fine. `done` was already called this way for other errors, so it should be fine. Even though behavior is different now.